File: js/hang/src/catalog/writer.ts
import type * as Moq from "@moq/lite";
import { type Effect, Signal } from "@moq/signals";
import type { Section } from "./section";

/// A catalog writer that manages typed sections and serializes them to a MoQ track.
///
/// Each section is a Signal that can be set independently. When served on a track,
/// changes are reactively detected and the full catalog JSON is re-serialized.
/// Microtask coalescing in Signal means multiple set() calls in the same tick
/// produce a single write.
export class CatalogWriter {
	// biome-ignore lint/suspicious/noExplicitAny: we store heterogeneous section types
	#sections = new Map<string, { signal: Signal<any> }>();

	/// Register a section for writing. Returns a Signal<T | undefined> for read+write.
	section<T>(def: Section<T>): Signal<T | undefined> {
		const existing = this.#sections.get(def.name);
		if (existing) return existing.signal as Signal<T | undefined>;

		const signal = new Signal<T | undefined>(undefined);
		this.#sections.set(def.name, { signal });
		return signal;
	}

	/// Serve the catalog on a MoQ track.
	///
	/// Uses Effect to reactively subscribe to all registered section signals.
	/// When any signal changes, re-serializes and writes a new frame.
	serve(track: Moq.Track, effect: Effect): void {
		effect.run((inner) => {
			const obj: Record<string, unknown> = {};

			for (const [name, { signal }] of this.#sections) {
				const value = inner.get(signal);
				if (value !== undefined) {
					obj[name] = value;
				}
			}

			const encoder = new TextEncoder();
			track.writeFrame(encoder.encode(JSON.stringify(obj)));
		});
	}
}

File: rs/hang/src/catalog/writer.rs
use std::collections::HashMap;

use serde::Serialize;

use crate::Result;

use super::Section;

/// Shared state for catalog sections.
#[derive(Default)]
pub struct CatalogState {
	pub sections: HashMap<String, serde_json::Value>,
}

/// A catalog writer that manages typed sections and serializes them to JSON.
///
/// Each section is identified by a name and stores a typed value.
/// The writer can encode all sections into a single JSON object for publishing.
#[derive(Clone)]
pub struct CatalogWriter {
	state: conducer::Producer<CatalogState>,
}

impl Default for CatalogWriter {
	fn default() -> Self {
		Self::new()
	}
}

impl CatalogWriter {
	pub fn new() -> Self {
		Self {
			state: conducer::Producer::new(CatalogState::default()),
		}
	}

	/// Write a section value. Serializes T to a JSON Value and stores it.
	/// Consumers are notified of the change.
	pub fn set<T: Serialize>(&self, section: &Section<T>, value: &T) -> Result<()> {
		let json = serde_json::to_value(value)?;
		let mut state = self.state.write().map_err(|_| crate::Error::Closed)?;
		state.sections.insert(section.name.to_string(), json);
		Ok(())
	}

	/// Remove a section from the catalog.
	/// Consumers are notified of the change.
	pub fn remove(&self, name: &str) -> Result<()> {
		let mut state = self.state.write().map_err(|_| crate::Error::Closed)?;
		state.sections.remove(name);
		Ok(())
	}

	/// Create a consumer that gets notified on any catalog change.
	pub fn consume(&self) -> conducer::Consumer<CatalogState> {
		self.state.consume()
	}

	/// Get read-only access to the current state.
	pub fn read(&self) -> conducer::Ref<'_, CatalogState> {
		self.state.read()
	}

	/// Serialize all sections to a JSON byte vector.
	pub fn encode(&self) -> Result<Vec<u8>> {
		let state = self.state.read();
		Ok(serde_json::to_vec(&state.sections)?)
	}

	/// Close the writer, notifying all consumers.
	pub fn close(&self) {
		let _ = self.state.close();
	}
}
